src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -79,30 +82,78 @@ public string CreateSignedHttpRequest(SignedHttpRequestDescriptor signedHttpRequ
         /// <returns>A signed http request as a JWS in Compact Serialization Format.</returns>
         public string CreateSignedHttpRequest(SignedHttpRequestDescriptor signedHttpRequestDescriptor, CallContext callContext)
         {
+            if (callContext != null)
+                LogHelper.LogVerbose(callContext.ActivityId.ToString());
+
             if (signedHttpRequestDescriptor == null)
                 throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor));
 
             if (signedHttpRequestDescriptor.SigningCredentials == null)
                 throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor.SigningCredentials));
 
-            var payload = CreateHttpRequestPayload(signedHttpRequestDescriptor, callContext);
-            var header = new JObject();
-            if (signedHttpRequestDescriptor.AdditionalHeaderClaims != null && signedHttpRequestDescriptor.AdditionalHeaderClaims.Count != 0)
+            string encodedPayload;
+            using (MemoryStream payloadUtf8Bytes = new MemoryStream())
             {
-                if (signedHttpRequestDescriptor.AdditionalHeaderClaims.Keys.Intersect(JwtTokenUtilities.DefaultHeaderParameters, StringComparer.OrdinalIgnoreCase).Any())
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenException(LogHelper.FormatInvariant(JsonWebTokens.LogMessages.IDX14116, LogHelper.MarkAsNonPII(nameof(signedHttpRequestDescriptor.AdditionalHeaderClaims)), LogHelper.MarkAsNonPII(string.Join(", ", JwtTokenUtilities.DefaultHeaderParameters)))));
 
-                header.Merge(JObject.FromObject(signedHttpRequestDescriptor.AdditionalHeaderClaims));
+                Utf8JsonWriter payloadWriter = null;
+                try
+                {
+                    payloadWriter = new Utf8JsonWriter(payloadUtf8Bytes, new JsonWriterOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping });
+                    payloadWriter.WriteStartObject();
+
+                    CreateHttpRequestPayload(ref payloadWriter, signedHttpRequestDescriptor);
+
+                    payloadWriter.WriteEndObject();
+                    payloadWriter.Flush();
+                    encodedPayload = Base64UrlEncoder.Encode(payloadUtf8Bytes.ToArray());
```

Comment:
same comment as above can we use the streams buffer?

Reply:
Yep.

Reply:
@keegan-caruso addressed in d789adbb30b8093a0eff70ee545983195935366f